src/AspNet.Security.OAuth.QQ/QQAuthenticationHandler.cs

```diff
@@ -0,0 +1,127 @@
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Security.Claims;
+using System.Threading.Tasks;
+using AspNet.Security.OAuth.Extensions;
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.OAuth;
+using Microsoft.AspNetCore.Http.Authentication;
+using Microsoft.AspNetCore.WebUtilities;
+using Microsoft.Extensions.Logging;
+using Newtonsoft.Json.Linq;
+using JetBrains.Annotations;
+
+namespace AspNet.Security.OAuth.QQ
+{
+    public class QQAuthenticationHandler : OAuthHandler<QQAuthenticationOptions>
+    {
+        public QQAuthenticationHandler([NotNull] HttpClient client)
+            : base(client)
+        {
+        }
+
+        protected override async Task<AuthenticationTicket> CreateTicketAsync(ClaimsIdentity identity, AuthenticationProperties properties, OAuthTokenResponse tokens)
+        {
+            var openid = await GetUserOpenIdAsync(tokens);
+            var queryString = new Dictionary<string, string>()
+            {
+                {"oauth_consumer_key",Options.ClientId },
+                {"access_token",tokens.AccessToken },
+                {"openid",openid },
+            };
+
+            var endpoint = QueryHelpers.AddQueryString(Options.UserInformationEndpoint, queryString);
+            var response = await Backchannel.GetAsync(endpoint);
+            if (!response.IsSuccessStatusCode)
+            {
+                Logger.LogError("An error occurred while retrieving the user profile: the remote server " +
+                                "returned a {Status} response with the following payload: {Headers} {Body}.",
+                                /* Status: */ response.StatusCode,
+                                /* Headers: */ response.Headers.ToString(),
+                                /* Body: */ await response.Content.ReadAsStringAsync());
+
+                throw new HttpRequestException("An error occurred when retrieving user information.");
+            }
+
+            var payload = JObject.Parse(await response.Content.ReadAsStringAsync());
+            var ret = payload.Value<int>("ret");
+            if (ret != 0)
+            {
+                var msg = payload.Value<string>("msg");
+                throw new HttpRequestException($"An error occurred when retrieving user information. code: {ret}, msg: {msg}");
+            }
+
+            identity.AddOptionalClaim(ClaimTypes.NameIdentifier, openid, Options.ClaimsIssuer)
+                    .AddOptionalClaim(ClaimTypes.Name, QQAuthenticationHelper.GetNickname(payload), Options.ClaimsIssuer)
+                    .AddOptionalClaim(ClaimTypes.Gender, QQAuthenticationHelper.GetGender(payload), Options.ClaimsIssuer)
+                    .AddOptionalClaim("urn:qq:figureurl", QQAuthenticationHelper.GetFigureUrl(payload), Options.ClaimsIssuer)
+                    .AddOptionalClaim("urn:qq:figureurl_1", QQAuthenticationHelper.GetFigureUrl1(payload), Options.ClaimsIssuer)
```

Comment:
Not sure I like these _1/_2/qq_1/qq_2 claims. Do we really need them? What do they do?